test/integration-tests/tests/jd_integration.rs

```diff
@@ -108,3 +114,130 @@ async fn jdc_does_not_stackoverflow_when_no_token() {
     tokio::time::sleep(Duration::from_secs(3)).await;
     assert!(tokio::net::TcpListener::bind(jdc_addr).await.is_err());
 }
+
+///The issue arises from the following message flow. JDC first receives a `NewTemplate`,
+/// prompting it to send a `RequestTransactionData` message to TP. TP responds with
+/// `RequestTransactionDataSuccess`, after which JDC sends a `DeclareMiningJob` to JDS. JDS then
+/// requests missing transactions from JDC via `ProvideMissingTransactions`. While JDC is in the
+/// process of responding with `ProvideMissingTransactionsSuccess`, a block is found, leading JDC
+/// to send a `SubmitSolution` to JDS. However, since JDS has not yet received the requested
+/// transactions from JDC, it results in shutdown. This test tries to mimic the given situation.
+#[tokio::test]
+async fn jds_receive_solution_while_processing_declared_job_test() {
+    start_tracing();
+    let (tp_1, tp_addr_1) = start_template_provider(None);
+
+    let (tp_2, tp_addr_2) = start_template_provider(None);
+
+    let (_pool, pool_addr) = start_pool(Some(tp_addr_1)).await;
+    let (_jds, jds_addr) = start_jds(tp_1.rpc_info()).await;
+
+    // Dummy data to construct a SubmitSolution message
+    let prev_hash = U256::Owned(vec![
+        184, 103, 138, 88, 153, 105, 236, 29, 123, 246, 107, 203, 1, 33, 10, 122, 188, 139, 218,
+        141, 62, 177, 158, 101, 125, 92, 214, 150, 199, 220, 29, 8,
+    ]);
+
+    let extranonce = B032::Owned(vec![
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0,
+        0, 0,
+    ]);
+
+    let submit_solution_replace = ReplaceMessage::new(
+        MessageDirection::ToUpstream,
+        MESSAGE_TYPE_PROVIDE_MISSING_TRANSACTIONS_SUCCESS,
+        AnyMessage::JobDeclaration(roles_logic_sv2::parsers::JobDeclaration::SubmitSolution(
+            SubmitSolutionJd {
+                ntime: 0,
+                nbits: 0,
+                nonce: 0,
+                version: 0,
+                prev_hash,
+                extranonce,
+            },
+        )),
+    );
+
+    // This sniffer sits between `jds` and `sniffer_b`, intercepting replaced `submit_solution`
+    // messages from downstream.
+    let (sniffer_a, sniffer_a_addr) = start_sniffer("A".to_string(), jds_addr, false, None).await;
+
+    // This sniffer sits between `sniffer_a` and `jdc`, replacing `ProvideMissingTransactionSuccess`
+    // with `SubmitSolution`.
+    let (sniffer_b, sniffer_b_addr) = start_sniffer(
+        "B".to_string(),
+        sniffer_a_addr,
+        false,
+        Some(submit_solution_replace.into()),
+    )
+    .await;
```

Comment:
I dont understand why you need two sniffers?

Reply:
no need to test this. we know how the sniffer behaves and we have tests specifically for that in sniffer_integration. you can get away with this test with a single sniffer.